### PR TITLE
Metrics: MetricPointValueStorage explicit layout struct part2

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Followed-up on [conversation](https://github.com/open-telemetry/opentelemetry-dotnet/pull/2694#issuecomment-981171861) from the last PR. Moved the histogram sum tracking to `HistogramBuckets`. That saved 8 bytes for the non-histogram `MetricPoint`s. Also used `SnapshotBucketCounts` for the histogram lock to save an `object` allocation.

* Added missing xml comments in `MetricPoint`, `HistogramBuckets`, & `HistogramBucket`.

* Added some inlining hints on `MetricPoint`.

* Renamed from "current"/"snapshot" to "running"/"snapshot" which seemed closer to what is going on.

### Before

|                     AggType | Allocated |
|---------------------------- |----------:|
|   LongSumIncomingCumulative |     112 B |
| DoubleSumIncomingCumulative |     112 B |
|                   Histogram |     408 B |

### After

|                     AggType | Allocated |
|---------------------------- |----------:|
|   LongSumIncomingCumulative |     104 B |
| DoubleSumIncomingCumulative |     104 B |
|                   Histogram |     384 B |

## Investigation Notes

What I really wanted to do was this (more or less):

```csharp
    [StructLayout(LayoutKind.Explicit)]
    internal struct MetricPointSecondaryValueStorage
    {
        [FieldOffset(0)]
        public double DeltaLastValue;

        [FieldOffset(0)]
        public HistogramBuckets HistogramBuckets;
    }
```

Basically use 8 bytes for either the delta value or the pointer to `HistogramBuckets`. Actually compiles fine, but blows up trying to load the assembly. Something about you can't overlap a reference object with a value type because it would mess up the GC's ability to track that reference in order to know if it should be cleaned up or not.

### CODE REVIEW

File: src/OpenTelemetry/Metrics/HistogramBuckets.cs [hunk 1]
```diff
@@ -16,26 +16,37 @@
...
+
+        internal HistogramBuckets(double[] explicitBounds)
         {
-            this.ExplicitBounds = histogramBounds;
-            this.BucketCounts = histogramBounds != null ? new long[histogramBounds.Length + 1] : null;
-            this.AggregatedBucketCounts = histogramBounds != null ? new long[histogramBounds.Length + 1] : null;
-            this.LockObject = new object();
+            this.ExplicitBounds = explicitBounds;
+            this.RunningBucketCounts = explicitBounds != null ? new long[explicitBounds.Length + 1] : null;
+            this.SnapshotBucketCounts = explicitBounds != null ? new long[explicitBounds.Length + 1] : new long[0];
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/HistogramBuckets.cs:39]: Use `Array.Empty<long>()`?

File: src/OpenTelemetry/Metrics/MetricPoint.cs [hunk 2]
```diff
@@ -21,37 +21,45 @@
...
     {
         private readonly AggregationType aggType;
+
         private readonly HistogramBuckets histogramBuckets;
 
         // Represents either "value" for double/long metric types or "count" when histogram
-        private MetricPointValueStorage primaryValue;
+        private MetricPointValueStorage runningValue;
+
+        // Represents either "value" for double/long metric types or "count" when histogram
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/MetricPoint.cs:36]:
```suggestion
        // Represents temporality adjusted "value" for double/long metric types or "count" when histogram
```

File: src/OpenTelemetry/Metrics/HistogramBuckets.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/HistogramBuckets.cs:39]: I'm using `SnapshotBucketCounts` for the lock now so it has to be something uniquely held by each instance of `HistogramBuckets`. The `Array.Empty<T>` instances are shared (AFAIK) so that might lead to some contention.

Comment (utpilla) [reply src/OpenTelemetry/Metrics/HistogramBuckets.cs:39]: Why not lock on `this.HistogramBuckets` in `MetricPoint?

File: src/OpenTelemetry/Metrics/MetricPoint.cs [hunk 3]
```diff
@@ -21,37 +21,45 @@
...
+        private MetricPointValueStorage snapshotValue;
 
-        // Represents either "lastValue" for double/long metric types when delta or "sum" when histogram
```
Comment (utpilla) [inline src/OpenTelemetry/Metrics/MetricPoint.cs:32]: We could still have this comment for this field.

**Review by utpilla — APPROVED** [review]

File: src/OpenTelemetry/Metrics/MetricPoint.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricPoint.cs:36]: Updated. Just to confirm, you intended for it to be "temporality " in the comment and not "temporarily" right?

File: src/OpenTelemetry/Metrics/HistogramBuckets.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/HistogramBuckets.cs:39]: We return the `HistogramBuckets` instance to users in the `MetricPoint.GetHistogramBuckets()` method. So they could (in theory) use that for their own locking purposes. That could lead to contention or (worse case) deadlock. I wanted to use something internal. Basically this concept of [weak identity vs strong identity](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2002).

File: src/OpenTelemetry/Metrics/MetricPoint.cs (hunk 3 above)
Comment (CodeBlanch) [reply src/OpenTelemetry/Metrics/MetricPoint.cs:32]: I can add it, but I don't think it is really needed. Because the field name on `HistogramBuckets` is `RunningSum`/`SnapshotSum` which more clearly conveys the (single) intention/purpose. The reason I have the comments on `MetricPoint` is the names `runningValue`/`snapshotValue` don't really indicate they also stand in for the "Count" storage. Those fields are multi-purpose.

File: src/OpenTelemetry/Metrics/MetricPoint.cs (hunk 2 above)
Comment (utpilla) [reply src/OpenTelemetry/Metrics/MetricPoint.cs:36]: Yes "Temporality" (Delta vs Cumulative).

**Comment by mic-max** [comment]
LGTM 👍

**Review by cijothomas — APPROVED** [review]
